nmmo/task/task_api.py

```diff
@@ -1,81 +1,146 @@
-from typing import Dict, List, Tuple
-
+from __future__ import annotations
+from typing import Dict, List, Tuple, TYPE_CHECKING
+from functools import reduce
+from abc import ABC, abstractmethod
 from pettingzoo.utils.env import AgentID
 
 import nmmo
 from nmmo.task.game_state import GameStateGenerator
 from nmmo.task.predicate import Predicate
+from nmmo.task.group import Group
+
+if TYPE_CHECKING:
+  from nmmo.task.game_state import GameState
+
+class Task(ABC):
+  def __init__(self, assignee: Group):
+    self._assignee = assignee
+
+  @property
+  def assignee(self):
+    return self._assignee
+
+  @abstractmethod
+  def rewards(self, gs: GameState) -> Tuple[Dict[int, float], Dict[int, Dict]]:
+    """ Returns a mapping from ent_id to rewards and infos for all 
+    entities in assignee
+    """
+    raise NotImplementedError
+
+class PredicateTask(Task, ABC):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate):
+    super().__init__(assignee)
+    self._predicate = predicate
+
+  def evaluate(self, gs: GameState) -> float:
+    name = self._predicate.name
+    cache = gs.cache_result
+    if name not in cache:
+      cache[name] = self._predicate(gs)
+    return cache[name]
+
+class Once(PredicateTask):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate,
+               reward = 1):
+    super().__init__(assignee, predicate)
+    self._reward = reward
+    self._completed = False
+
+  def rewards(self, gs: GameState):
+    rewards = {int(ent_id): 0 for ent_id in self._assignee}
+    infos = {int(ent_id): {self._predicate.name: self.evaluate(gs)}
+             for ent_id in self._assignee}
+    if not self._completed and self.evaluate(gs):
+      self._completed = True
+      rewards = {int(ent_id): self._reward for ent_id in self._assignee}
+    return rewards, infos
+
+class Repeat(PredicateTask):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate,
+               reward = 1):
+    super().__init__(assignee, predicate)
+    self._reward = reward
+
+  def rewards(self, gs: GameState):
+    rewards = {int(ent_id): 0 for ent_id in self._assignee}
+    infos = {int(ent_id): {self._predicate.name: self.evaluate(gs)}
+             for ent_id in self._assignee}
+    if self.evaluate(gs):
+      rewards = {int(ent_id): self._reward for ent_id in self._assignee}
+    return rewards, infos
 
+class MultiTask(Task):
+  def __init__(self, *tasks: Task):
```

Comment:
Might make sense to allow you to weight tasks, so provide a list of task weights or make tasks a dictionary of {task:weight}